calicoctl/commands/migrate/import.go

```diff
@@ -0,0 +1,409 @@
+			"<NAME>":   []string{},
+			"--config": args["--config"].(string),
+			"--export": false,
+			"--output": "ps",
+			"get":      true,
+		}
+
+		if _, ok := namespacedResources[r]; ok {
+			mockArgs["--all-namespaces"] = true
+		}
+
+		// Get resources
+		results := common.ExecuteConfigCommand(mockArgs, common.ActionGetOrList)
+
+		// Loop through the result lists and see if anything exists
+		for _, resource := range results.Resources {
+			if meta.LenList(resource) > 0 {
+				return fmt.Errorf("Found existing Calico %s resource", results.SingleKind)
+			}
+
+			if results.FileInvalid {
+				return fmt.Errorf("Failed to execute command: %v", results.Err)
+			} else if results.Err != nil {
+				return fmt.Errorf("Failed to retrieve %s resources during datastore check: %v", resourceDisplayMap[r], results.Err)
+			}
+		}
+	}
+
+	// Check if any IPAM resources exist
+	ipam := NewMigrateIPAM(c)
+	err := ipam.PullFromDatastore()
+	if err != nil {
+		return fmt.Errorf("Failed to retrieve IPAM resources during datastore check: %s", err)
+	}
+
+	if !ipam.IsEmpty() {
+		return fmt.Errorf("Found existing IPAM resources")
+	}
+
+	return nil
+}
+
+func updateClusterInfo(ctx context.Context, c client.Interface, clusterInfoJson []byte) error {
+	// Unmarshal the etcd cluster info resource.
+	migrated := apiv3.ClusterInformation{}
+	err := json.Unmarshal(clusterInfoJson, &migrated)
+	if err != nil {
+		return fmt.Errorf("Error reading exported cluster info for migration: %s", err)
+	}
+
+	// Get the "default" cluster info resource.
+	clusterinfo, err := c.ClusterInformation().Get(ctx, "default", options.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("Error retrieving current cluster info for migration: %s", err)
+	}
+
+	// Update the calico version and cluster GUID.
+	clusterinfo.Spec.ClusterGUID = migrated.Spec.ClusterGUID
+	clusterinfo.Spec.CalicoVersion = migrated.Spec.CalicoVersion
+	_, err = c.ClusterInformation().Update(ctx, clusterinfo, options.SetOptions{})
+	if err != nil {
+		return fmt.Errorf("Error updating current cluster info for migration: %s", err)
+	}
+
+	return nil
+}
+
+func updateV3Resources(cf string, data []byte) error {
+	// Create tempfile so the v3 resources can be created using Apply
+	tempfile, err := ioutil.TempFile("", "v3migration")
+	if err != nil {
+		return fmt.Errorf("Error while creating temporary v3 migration file: %s\n", err)
+	}
+	defer os.Remove(tempfile.Name())
+
+	if _, err := tempfile.Write(data); err != nil {
+		return fmt.Errorf("Error while writing to temporary v3 migration file: %s\n", err)
+	}
+
+	mockArgs := map[string]interface{}{
+		"--config":   cf,
+		"--filename": tempfile.Name(),
+		"apply":      true,
+	}
+	err = applyV3(mockArgs)
+	if err != nil {
+		return fmt.Errorf("Failed to import v3 resources: %s\n", err)
+	}
+
+	return nil
+}
+
+func importCRDs(cfg *apiconfig.CalicoAPIConfig) error {
+	// TODO: Add the apiextensions clientset and handling logic to the libcalico-go kubeclient
```

Comment:
I don't think we want to do this TODO.

Reply:
I thought it would be nicer to not have what is essentially client level code living in calicoctl since it feels like it would get more use in libcalico-go. Granted, we don't have a use case for adding it yet, but thought it would be a nice thing to have.

Reply:
I'd prefer to put code closer to where it's used - nothing in libcalico uses apiextensions.